File: src/backend/trip-service/app/services/driver_service.py
"""
DynamoDB CRUD for drivers.

Table:  {env}-driver-assistant-drivers
PK:     phone  (e.g. +84867347452)
"""
from datetime import datetime, timezone
from functools import lru_cache

import boto3
from boto3.dynamodb.conditions import Key

from app.core.config import get_settings


@lru_cache
def _table():
    s = get_settings()
    dyn = boto3.resource("dynamodb", region_name="ap-southeast-1")
    drivers_table = s.dynamodb_table.replace("-trips", "-drivers")
    return dyn.Table(drivers_table)


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def register(phone: str, name: str, fcm_token: str = "") -> dict:
    driver = {
        "phone":         phone,
        "name":          name,
        "fcm_token":     fcm_token,
        "status":        "offline",
        "registered_at": _now(),
        "updated_at":    _now(),
    }
    _table().put_item(Item=driver)
    return driver


def get(phone: str) -> dict | None:
    resp = _table().get_item(Key={"phone": phone})
    return resp.get("Item")


def update(phone: str, fields: dict) -> dict | None:
    fields["updated_at"] = _now()
    names  = {f"#{k}": k for k in fields}
    values = {f":{k}": v for k, v in fields.items()}
    expr   = "SET " + ", ".join(f"#{k} = :{k}" for k in fields)

    resp = _table().update_item(
        Key={"phone": phone},
        UpdateExpression=expr,
        ExpressionAttributeNames=names,
        ExpressionAttributeValues=values,
        ReturnValues="ALL_NEW",
    )
    return resp.get("Attributes")


def list_all() -> list[dict]:
    resp = _table().scan()
    items = resp.get("Items", [])
    return sorted(items, key=lambda x: x.get("name", ""))


def list_online() -> list[dict]:
    return [d for d in list_all() if d.get("status") == "online"]

File: src/backend/trip-service/src/services/driverService.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  PutCommand,
  GetCommand,
  ScanCommand,
  UpdateCommand,
} from '@aws-sdk/lib-dynamodb';
import { config } from '../config';

const driversTable = config.dynamodbTable.replace('-trips', '-drivers');
const docClient = DynamoDBDocumentClient.from(
  new DynamoDBClient({ region: config.awsRegion })
);

function now(): string {
  return new Date().toISOString();
}

export interface Driver {
  phone: string;
  name: string;
  fcm_token: string;
  status: string;
  registered_at: string;
  updated_at: string;
}

export async function register(phone: string, name: string, fcmToken = ''): Promise<Driver> {
  const driver: Driver = {
    phone, name, fcm_token: fcmToken,
    status: 'offline',
    registered_at: now(),
    updated_at: now(),
  };
  await docClient.send(new PutCommand({ TableName: driversTable, Item: driver }));
  return driver;
}

export async function get(phone: string): Promise<Driver | null> {
  const result = await docClient.send(new GetCommand({
    TableName: driversTable,
    Key: { phone },
  }));
  return (result.Item as Driver) ?? null;
}

export async function update(phone: string, fields: Record<string, unknown>): Promise<Driver | null> {
  const data: Record<string, unknown> = { ...fields, updated_at: now() };
  const names: Record<string, string> = {};
  const values: Record<string, unknown> = {};
  for (const k of Object.keys(data)) {
    names[`#${k}`] = k;
    values[`:${k}`] = data[k];
  }
  const setExpr = 'SET ' + Object.keys(data).map(k => `#${k} = :${k}`).join(', ');

  const result = await docClient.send(new UpdateCommand({
    TableName: driversTable,
    Key: { phone },
    UpdateExpression: setExpr,
    ExpressionAttributeNames: names,
    ExpressionAttributeValues: values,
    ReturnValues: 'ALL_NEW',
  }));
  return (result.Attributes as Driver) ?? null;
}

export async function listAll(): Promise<Driver[]> {
  const result = await docClient.send(new ScanCommand({ TableName: driversTable }));
  const items = (result.Items as Driver[]) ?? [];
  return items.sort((a, b) => a.name.localeCompare(b.name));
}

export function listOnline(drivers: Driver[]): Driver[] {
  return drivers.filter(d => d.status === 'online');
}
